components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/nats/CoreListener.java

```diff
@@ -0,0 +1,270 @@
+        String traceConnection = natsProperties.getProperty(NatsConstants.TRACE_CONNECTION);
+        String tlsProtocol = validateParameter(natsProperties.getProperty(NatsConstants.TLS_PROTOCOL));
+        String tlsKeyStoreType = validateParameter(natsProperties.getProperty(NatsConstants.TLS_KEYSTORE_TYPE));
+        String tlsKeyStoreLocation = validateParameter(natsProperties.getProperty(NatsConstants.TLS_KEYSTORE_LOCATION));
+        String tlsKeyStorePassword = validateParameter(natsProperties.getProperty(NatsConstants.TLS_KEYSTORE_PASSWORD));
+        String tlsTrustStoreType = validateParameter(natsProperties.getProperty(NatsConstants.TLS_TRUSTSTORE_TYPE));
+        String tlsTrustStoreLocation = validateParameter(natsProperties.getProperty(NatsConstants.TLS_TRUSTSTORE_LOCATION));
+        String tlsTrustStorePassword = validateParameter(natsProperties.getProperty(NatsConstants.TLS_TRUSTSTORE_PASSWORD));
+        String tlsKeyManagerAlgorithm = validateParameter(natsProperties.getProperty(NatsConstants.TLS_KEY_MANAGER_ALGORITHM));
+        String tlsTrustManagerAlgorithm = validateParameter(natsProperties.getProperty(NatsConstants.TLS_TRUST_MANAGER_ALGORITHM));
+
+        Options.Builder builder = new Options.Builder(natsProperties);
+
+        if (StringUtils.isNotEmpty(bufferSize)) {
+            builder.bufferSize(Integer.parseInt(bufferSize));
+        }
+
+        if (Boolean.parseBoolean(turnOnAdvancedStats)) {
+            builder.turnOnAdvancedStats();
+        }
+
+        if (Boolean.parseBoolean(traceConnection)) {
+            builder.traceConnection();
+        }
+
+        if (StringUtils.isNotEmpty(
+                tlsProtocol + tlsTrustStoreType + tlsTrustStoreLocation + tlsTrustStorePassword + tlsKeyStoreType
+                        + tlsKeyStoreLocation + tlsKeyStorePassword + tlsKeyManagerAlgorithm
+                        + tlsTrustManagerAlgorithm)) {
+            SSLContext sslContext = createSSLContext(
+                    new TLSConnection(tlsProtocol, tlsTrustStoreType, tlsTrustStoreLocation, tlsTrustStorePassword,
+                            tlsKeyStoreType, tlsKeyStoreLocation, tlsKeyStorePassword, tlsKeyManagerAlgorithm,
+                            tlsTrustManagerAlgorithm));
+            if (sslContext != null) {
+                builder.sslContext(sslContext);
+            }
+        }
+        return Nats.connect(builder.build());
+    }
+
+    @Override public void consumeMessage(String sequenceName) throws InterruptedException {
+        Subscription subscription;
+        String queueGroup = natsProperties.getProperty(NatsConstants.QUEUE_GROUP);
+        if (StringUtils.isNotEmpty(queueGroup)) {
+            subscription = connection.subscribe(subject, queueGroup);
+        } else {
+            subscription = connection.subscribe(subject);
+        }
+        Message natsMessage = subscription.nextMessage(Duration.ZERO);
```

Comment:
As per API documentation at [1] `0` means infinite waiting. If there is no messages in the subject, this method will never return with `Duration=0`. We need to have something like 100ms. Then if there are no messages addressed to the subject, it will return null.